libvast/vast/command.hpp

```diff
@@ -0,0 +1,222 @@
+/******************************************************************************
+ *                    _   _____   __________                                  *
+ *                   | | / / _ | / __/_  __/     Visibility                   *
+ *                   | |/ / __ |_\ \  / /          Across                     *
+ *                   |___/_/ |_/___/ /_/       Space and Time                 *
+ *                                                                            *
+ * This file is part of VAST. It is subject to the license terms in the       *
+ * LICENSE file found in the top-level directory of this distribution and at  *
+ * http://vast.io/license. No part of VAST, including this file, may be       *
+ * copied, modified, propagated, or distributed except according to the terms *
+ * contained in the LICENSE file.                                             *
+ ******************************************************************************/
+
+#ifndef VAST_COMMAND_HPP
+#define VAST_COMMAND_HPP
+
+#include <functional>
+#include <memory>
+#include <string>
+#include <string_view>
+
+#include <caf/actor_system_config.hpp>
+#include <caf/fwd.hpp>
+#include <caf/message.hpp>
+
+#include "vast/error.hpp"
+#include "vast/data.hpp"
+
+#include "vast/concept/parseable/to.hpp"
+#include "vast/concept/parseable/vast/data.hpp"
+
+#include "vast/detail/steady_map.hpp"
+#include "vast/detail/string.hpp"
+
+namespace vast {
+
+/// A top-level command.
+class command {
+public:
+  // -- member types -----------------------------------------------------------
+
+  /// An option of a command.
+  struct option {
+    template <class T = bool>
+    static std::pair<std::string, option>
+    make(const std::string& tag, std::string desc, T x = {}) {
+      auto s = detail::split_to_str(tag, ",");
+      std::string shortcut;
+      if (s.size() >= 2)
+        shortcut = s[1][0];
+      return {s[0], {std::move(shortcut), std::move(desc),
+                    data{std::forward<T>(x)}}};
+    }
+
+    std::string shortcut;
+    std::string description;
+    data value;
```

Comment:
It looks like you no longer use the `value` member, but instead use `add_opt` with an lvalue. Why not store the option value directly where it belongs? The DSL I came up with looked like this:

```cpp
cmd("foo")
  .opt<std::string>(..., "default")
  .opt<double>(..., 4.2)
  .opt<address>(*to<address>("192.168.10.1"));
```

There was also an accessor API along the lines of:

```cpp
auto opt = cmd["opt"]; // get a pointer to the option or nullptr if the option doesn't exist.
auto x = cmd->get("opt"); // get the option value or nullptr if the option doesn't exist.
```

Why did you prefer a design where the value is stored externally? I'm not opposed to it, just wondering, since both styles have advantages and disadvantages.

Reply:
Since all of this builds on top of CAF's CLI parsing mechanisms, I found building another layer on top of that adds too much complexity. The current approach only leverages CAF building blocks without implementing a separate DSL. Storing values externally is simply a consequence of using CAF's parser directly.